### PR TITLE
feat: Add AWS::ElasticBeanstalk::ApplicationVersion

### LINKED ISSUE #621: Ensure Support for AWS::ElasticBeanstalk::*
- [ ] AWS::ElasticBeanstalk::Application
- [ ] AWS::ElasticBeanstalk::ApplicationVersion
- [ ] AWS::ElasticBeanstalk::Environment

### PR DESCRIPTION
**Pull request by erezrokah**
Fixes https://github.com/cloudquery/cq-provider-aws/issues/621

Ensured existing logic works, and added support for `AWS::ElasticBeanstalk::ApplicationVersion`

## Todo
- [x] Handle the bug reported here https://github.com/cloudquery/cq-provider-aws/issues/621#issuecomment-1103846841

## Other considerations
- [ ] Should we create a relation from applications to application versions?

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -7,14 +7,14 @@ install-cq:
 	chmod a+x cloudquery
 
 # start a timescale db running in a local container
-.PHONY: ts-start
-ts-start:
+.PHONY: start-ts
```
Comment (erezrokah) [inline Makefile:10]: Aligned the naming convention with postgres (e.g. `start-pg`)

File: Makefile [hunk 2]
```diff
@@ -7,14 +7,14 @@ install-cq:
...
+start-ts:
 	docker run -p 5433:5432 -e POSTGRES_PASSWORD=pass -d timescale/timescaledb:latest-pg14
 
 # stop the timescale db running in a local container
-.PHONY: ts-stop
-ts-stop:
-	docker stop $(docker ps -q --filter ancestor=timescale/timescaledb:latest-pg14)
+.PHONY: stop-ts
+stop-ts:
+	docker stop $$(docker ps -q --filter ancestor=timescale/timescaledb:latest-pg14)
```
Comment (erezrokah) [inline Makefile:17]: This script was failing without the additional `$`

File: resources/provider/provider.go [hunk 3]
```diff
@@ -151,6 +151,7 @@ func Provider() *provider.Provider {
 			"efs.filesystems":                       efs.EfsFilesystems(),
 			"eks.clusters":                          eks.EksClusters(),
 			"elasticbeanstalk.applications":         elasticbeanstalk.ElasticbeanstalkApplications(),
+			"elasticbeanstalk.application_versions": elasticbeanstalk.ApplicationVersions(),
```
Comment (erezrokah) [inline resources/provider/provider.go:154]: The name without the prefix `elasticbeanstalk.ApplicationVersions` vs `elasticbeanstalk.ElasticbeanstalkApplicationVersions` was generated by `cq-gen`, so I kept it this way.

Happy to rename it

File: resources/services/elasticbeanstalk/application_versions.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+	for {
+		output, err := svc.DescribeApplicationVersions(ctx, &config, func(options *elasticbeanstalk.Options) {
+			options.Region = c.Region
+		})
+		if err != nil {
+			return diag.WrapError(err)
+		}
+
+		for i := range output.ApplicationVersions {
+			if output.ApplicationVersions[i].BuildArn == nil {
```
Comment (erezrokah) [inline resources/services/elasticbeanstalk/application_versions.go:129]: These `nil` values were failing the integration tests.

Please forgive my rusty `go` if I didn't follow a convention here

File: resources/services/elasticbeanstalk/gen.hcl [hunk 5]
```diff
@@ -0,0 +1,37 @@
+service          = "aws"
+output_directory = "."
+add_generate     = true
+
+resource "aws" "elasticbeanstalk" "application_versions" {
```
Comment (erezrokah) [inline resources/services/elasticbeanstalk/gen.hcl:5]: Added only the `application_versions` one. Not sure there's a reason to add others for existing code

File: Makefile [hunk 6]
```diff
@@ -7,14 +7,14 @@ install-cq:
 	chmod a+x cloudquery
```
Comment (erezrokah) [inline Makefile:7]: Changes to this files were extracted to a separate PR https://github.com/cloudquery/cq-provider-aws/pull/763

File: resources/services/elasticbeanstalk/environments.go [hunk 7]
```diff
@@ -433,6 +435,12 @@ func resolveElasticbeanstalkEnvironmentListeners(ctx context.Context, meta schem
 		ResourceArn: p.EnvironmentArn,
 	}, func(o *elasticbeanstalk.Options) {})
 	if err != nil {
+		// It takes a few minutes for an environment to be terminated
```
Comment (erezrokah) [inline resources/services/elasticbeanstalk/environments.go:438]: Fix for https://github.com/cloudquery/cq-provider-aws/issues/621#issuecomment-1103846841

File: resources/services/elasticbeanstalk/environments.go [hunk 8]
```diff
@@ -450,6 +458,17 @@ func fetchElasticbeanstalkEnvironmentLinks(ctx context.Context, meta schema.Clie
 	return nil
 }
 
+// It takes a few minutes for an environment to be terminated
+// This ensures we don't error while trying to fetch related resources for a terminated environment
+func IsInvalidParameterValueError(err error) bool {
```
Comment (erezrokah) [inline resources/services/elasticbeanstalk/environments.go:463]: Fix for https://github.com/cloudquery/cq-provider-aws/issues/621#issuecomment-1103846841

File: resources/services/elasticbeanstalk/environments.go (hunk 7 above)
Comment (disq) [reply resources/services/elasticbeanstalk/environments.go:438]: Please use `c.IsNotFoundError(err)` instead

Comment (erezrokah) [reply resources/services/elasticbeanstalk/environments.go:439]: Thanks for the tip! Done in https://github.com/cloudquery/cq-provider-aws/pull/751/commits/28ffc78101822538c840b3f7cc59a7fffc6346f3

File: Makefile [hunk 9]
```diff
@@ -64,6 +64,6 @@ test-integration:
 # Create a DB migration
 .PHONY: db-migration
 db-migration:
-	@if [[ "$(prefix)" == "" ]]; then echo "Invalid prefix, see example 'make db-migration prefix=26_v0.10.18'" && exit 1; fi;
+	@if [[ "$(prefix)" == "" ]]; then echo "Invalid prefix, see example 'make db-migration prefix=27_v0.10.20'" && exit 1; fi;
```
Comment (erezrokah) [inline Makefile:67]: Updated the script to use a prefix that we actually have (after merging this PR)

Comment (bbernays) [reply Makefile:67]: There is no reason why I think we can't autogenerate the prefix, or at least propose a reasonable option.


Here is a bash one liner that grabs the latest release (for the repo the command is run in) then increments the patch version and stores it in the `nextVersion` variable
```bash
nextVersion=$(gh release view --json tagName --jq .tagName | awk -F. -v OFS=. '{$NF += 1 ; print}')
```

Comment (erezrokah) [reply Makefile:67]: Great suggestion! What do you think about using `git tag | sort -V | tail -1` instead of `gh release view`?
The latter requires having the GitHub CLI installed and authenticated. The former requires having recent tags fetched from the GitHub repository. The former might fail more gracefully as at least some tags should be available

Comment (erezrokah) [reply Makefile:67]: Extracted this into a separate PR https://github.com/cloudquery/cq-provider-aws/pull/794

File: resources/services/elasticbeanstalk/application_versions.go [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+			{
+				Name:        "application_name",
+				Description: "The name of the application to which the application version belongs.",
+				Type:        schema.TypeString,
+			},
+			{
+				Name:        "application_version_arn",
+				Description: "The Amazon Resource Name (ARN) of the application version.",
+				Type:        schema.TypeString,
+			},
```
Comment (roneli) [inline resources/services/elasticbeanstalk/application_versions.go:46]: would change to "arn"

File: resources/services/elasticbeanstalk/application_versions.go [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+				Description: "Location where the repository is stored.  * CodeCommit  * S3  This member is required.",
+				Type:        schema.TypeString,
+				Resolver:    schema.PathResolver("SourceBuildInformation.SourceRepository"),
+			},
+			{
+				Name:        "source_build_information_source_type",
+				Description: "The type of repository.  * Git  * Zip  This member is required.",
+				Type:        schema.TypeString,
+				Resolver:    schema.PathResolver("SourceBuildInformation.SourceType"),
+			},
```
Comment (roneli) [inline resources/services/elasticbeanstalk/application_versions.go:84]: maybe remove the source_build_information prefix? cc @bbernays

File: resources/services/elasticbeanstalk/application_versions.go (hunk 4 above)
Comment (roneli) [reply resources/services/elasticbeanstalk/application_versions.go:129]: what you would do here is add an IgnoreInTests to the column

File: resources/services/elasticbeanstalk/environments.go (hunk 8 above)
Comment (roneli) [inline resources/services/elasticbeanstalk/environments.go:463]: maybe move this into a more generic place i.e client, as I am sure more resources might want to use this method.

File: resources/services/elasticbeanstalk/environments.go [hunk 12]
```diff
@@ -462,6 +481,9 @@ func fetchElasticbeanstalkConfigurationOptions(ctx context.Context, meta schema.
 		options.Region = c.Region
 	})
 	if err != nil {
+		if IsInvalidParameterValueError(err) {
+			return nil
```
Comment (roneli) [inline resources/services/elasticbeanstalk/environments.go:485]: debug log please

**Review by roneli — CHANGES_REQUESTED** [review]
Looks good, had some review comments

**Review by disq — COMMENTED** [review]
would also update the migration filename to 0.11.2

File: resources/services/elasticbeanstalk/environments.go (hunk 12 above)
Comment (erezrokah) [reply resources/services/elasticbeanstalk/environments.go:479]: Great point! Added in https://github.com/cloudquery/cq-provider-aws/pull/751/files#diff-e4759eab9747ef3abc27af2730ff5e7c5e63f486650f5678f85b8441a8db3a52R459

File: resources/services/elasticbeanstalk/application_versions.go (hunk 10 above)
Comment (erezrokah) [reply resources/services/elasticbeanstalk/application_versions.go]: Makes sense, renamed the field and fixed the migrations

File: resources/services/elasticbeanstalk/environments.go (hunk 8 above)
Comment (erezrokah) [reply resources/services/elasticbeanstalk/environments.go]: Done in https://github.com/cloudquery/cq-provider-aws/pull/751/commits/e17f2bf671a0d43a2bdccca2b6a3b828b83a40e1

File: resources/services/elasticbeanstalk/environments.go (hunk 12 above)
Comment (erezrokah) [reply resources/services/elasticbeanstalk/environments.go:479]: Done!

File: client/helpers.go [hunk 13]
```diff
@@ -278,6 +278,15 @@ func (c *Client) IsNotFoundError(err error) bool {
 	return false
 }
 
+func (c *Client) IsInvalidParameterValueError(err error) bool {
```
Comment (disq) [inline client/helpers.go:281]: We don't need `func (c *Client)`

**Review by erezrokah — COMMENTED** [review]
Thanks @roneli and @disq for the review.

I think I addressed all comments expect the naming for `source_build_information_source_location` (waiting on @bbernays).

Please let me know if I'm missing anything else

**Review by disq — APPROVED** [review]

File: resources/services/elasticbeanstalk/environments.go [hunk 14]
```diff
@@ -450,6 +456,10 @@ func fetchElasticbeanstalkEnvironmentLinks(ctx context.Context, meta schema.Clie
 	return nil
 }
 
+func logInvalidParameterValueError(meta schema.ClientMeta, err error, p types.EnvironmentDescription) {
+	meta.Logger().Debug("Failed extracting configuration settings for environment. It might be terminated", "environment", p.EnvironmentName, "application", p.ApplicationName)
```
Comment (disq) [inline resources/services/elasticbeanstalk/environments.go:460]: configuruation settings or configuration options? maybe remove the fn and have two separate log lines in the respective call sites

**Review by disq — CHANGES_REQUESTED** [review]

File: resources/services/elasticbeanstalk/environments.go (hunk 14 above)
Comment (erezrokah) [reply resources/services/elasticbeanstalk/environments.go:460]: Ah, good catch. Let me fix it

Comment (erezrokah) [reply resources/services/elasticbeanstalk/environments.go]: Done in https://github.com/cloudquery/cq-provider-aws/pull/751/commits/5cca3d4b154022366b35ac9d30f2d745dde702ff

**Review by disq — APPROVED** [review]

File: resources/services/elasticbeanstalk/application_versions.go [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+				Name:        "source_build_information_source_repository",
+				Description: "Location where the repository is stored.  * CodeCommit  * S3  This member is required.",
+				Type:        schema.TypeString,
+				Resolver:    schema.PathResolver("SourceBuildInformation.SourceRepository"),
+			},
+			{
+				Name:        "source_build_information_source_type",
+				Description: "The type of repository.  * Git  * Zip  This member is required.",
+				Type:        schema.TypeString,
+				Resolver:    schema.PathResolver("SourceBuildInformation.SourceType"),
```
Comment (roneli) [inline resources/services/elasticbeanstalk/application_versions.go:85]: nit: would remove the prefix here

**Review by roneli — APPROVED** [review]
LGTM, had one nit and we can merge this.

File: resources/services/elasticbeanstalk/application_versions.go (hunk 15 above)
Comment (erezrokah) [reply resources/services/elasticbeanstalk/application_versions.go]: Good point! Done in https://github.com/cloudquery/cq-provider-aws/pull/751/commits/b2d0102889ee8376bae07fa241cda78e406fc051
